### PR TITLE
feat: initialize typescript

### PR DESCRIPTION
**Pull request by johncardiologs**
A first jab at introducing ts/tsx to this repo!
cc [@jhash](https://github.com/jhash)

### CODE REVIEW

File: superglue/lib/components/Nav.tsx [hunk 1]
```diff
@@ -211,13 +229,14 @@ class Nav extends React.Component {
...
-}
+// TODO: remove now that we have Props ?
+// Nav.propTypes = {
+//   store: object,
+//   history: object,
+//   mapping: object,
+//   visit: func,
+//   remote: func,
+//   initialPageKey: string,
+// }
```
Comment (jho406) [inline superglue/lib/components/Nav.tsx:240]: Yes, i think it makes sense to remove this.

File: superglue/package.json [hunk 2]
```diff
@@ -56,6 +56,7 @@
     "redux-thunk": ">=2.3"
   },
   "dependencies": {
+    "@types/url-parse": "^1.4.11",
```
Comment (jho406) [inline superglue/package.json:59]: How do you feel about placing this into devDependencies? I don't plan to export types until superglue has types to export. So i prefer a smaller step.

File: superglue/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "jsx": "react",
+    "outDir": "./built",
```
Comment (jho406) [inline superglue/tsconfig.json:4]: I would like to continue using esbuild for the build process (for now as a microstep) and use typescript with noEmit. Could we remove line 4 for now?

File: superglue/package.json (hunk 2 above)
Comment (jho406) [reply superglue/package.json:59]: In this file, could we modify the scripts to something like the below:

```diff
-    "build:index": "babel lib -d dist",
-    "build": "npm run clean && npm run build:index && npm run copy:package && npm run copy:readme",
+    "build": "npm run clean && npm run build:js && npm run copy:package && npm run copy:readme",
+    "build:js": "babel lib --out-dir dist --extensions \".ts,.tsx\" --source-maps inline",
+    "type-check": "tsc --noEmit",
+    "type-check:watch": "npm run type-check -- --watch",
```

This goes with the theme of just setting up typescript on the newly created `ts` files, while allowing it to build despite the errors we see.

Comment (jho406) [reply superglue/package.json:59]: You'll also need to add `@babel/preset-typescript`, and make a modification to tsconfig. Something like

```
module.exports = {
  plugins: [
    ["transform-react-remove-prop-types", {removeImport: true}]
  ],
  presets: [
    [
      "@babel/preset-env",
      {
        useBuiltIns: "entry",
        corejs: "2",
        loose: true,
      }
    ],
    [
      "@babel/preset-react"
    ],
    [
      "@babel/preset-typescript"
    ]
  ],
};

```

**Review by jho406 — COMMENTED** [review]
I think this is looking great. Thank you so much. 

I want this PR to be sort of a smaller step, so i made comments to keep the esbuild process while allowing typescript to do checking instead of outputting JS. If you make those smaller changes, i'll merge this. Thanks again!

**Comment by johncardiologs** [comment]
Hi! Sorry, my mind's been completely elsewhere lately and I left this.

Agreed with the approach of keeping babel -- I made the changes you suggested.
I'll look into the failing tests & trying to get this to compile properly when I next have a chance!

File: superglue/package.json [hunk 4]
```diff
@@ -9,7 +9,10 @@
     "copy:package": "cat ./package.json | jq 'del(.scripts)' > dist/package.json",
     "copy:readme": "cp ../README.md dist/",
     "build:index": "babel lib -d dist",
```
Comment (jho406) [inline superglue/package.json:11]:
```suggestion
```

Comment (jho406) [reply superglue/package.json:11]: Lets remove this line since `build` no longer uses it.

**Review by jho406 — COMMENTED** [review]
Looks great! Just a few smaller comments. Don't worry about the typechecking errors. We can work on that on another issue. For this first step, I want this merged in so other contributors can start contributing in typescript. 

We just have to make sure that the build output is the same.

**Comment by jho406** [comment]
A rebase onto main would fix those errors.

File: superglue/package.json (hunk 4 above)
Comment (johncardiologs) [reply superglue/package.json]: Done!

**Comment by johncardiologs** [comment]
To compare the build outputs, I ran the builds and `diff`d the files. Primarily, I saw one added line in the typescript babel build outputs that was appended to the end of every file:
```
//# sourceMappingURL=data:application/json;charset=utf-8;base64
```
followed by a string of characters. It seems this corresponds to [babel's source map option](https://babeljs.io/docs/options#source-map-options).

The only other 'true diff' I saw apart from this was the following, where `dist/` contains files built via this PR and `maindist/` holds files built using the `npm run build:index` command from branch `main`:

```sh
❯ diff dist/index.js maindist/index.js
4c4
< exports.superglueReducer = exports.pageReducer = exports.mapDispatchToPropsIncludingVisitAndRemote = exports.getIn = exports.fragmentMiddleware = exports.UPDATE_FRAGMENTS = exports.SAVE_RESPONSE = exports.REMOVE_PAGE = exports.GRAFTING_SUCCESS = exports.GRAFTING_ERROR = exports.COPY_PAGE = exports.BEFORE_VISIT = exports.BEFORE_REMOTE = exports.BEFORE_FETCH = exports.ApplicationBase = void 0;
---
> exports.updateFragments = exports.superglueReducer = exports.pageReducer = exports.mapDispatchToPropsIncludingVisitAndRemote = exports.getIn = exports.fragmentMiddleware = exports.UPDATE_FRAGMENTS = exports.SAVE_RESPONSE = exports.REMOVE_PAGE = exports.GRAFTING_SUCCESS = exports.GRAFTING_ERROR = exports.COPY_PAGE = exports.BEFORE_VISIT = exports.BEFORE_REMOTE = exports.BEFORE_FETCH = exports.ApplicationBase = void 0;
10a11
> exports.updateFragments = _reducers.updateFragments;
```
I'm not entirely sure what could be causing this difference, but it doesn't seem super harmful to me -- though I'm not entirely familiar yet with superglue!

File: superglue/package.json [hunk 5]
```diff
@@ -8,8 +8,10 @@
     "clean": "rm -rf ./dist",
     "copy:package": "cat ./package.json | jq 'del(.scripts)' > dist/package.json",
     "copy:readme": "cp ../README.md dist/",
-    "build:index": "babel lib -d dist",
-    "build": "npm run clean && npm run build:index && npm run copy:package && npm run copy:readme",
+    "build": "npm run clean && npm run build:js && npm run copy:package && npm run copy:readme",
+    "build:js": "babel lib --out-dir dist --extensions \".ts,.tsx\" --source-maps inline",
```
Comment (jho406) [inline superglue/package.json:12]: Good eye with the source maps. To keep parity with main, lets remove the source maps option for now.

```suggestion
    "build:js": "babel lib --out-dir dist --extensions \".ts,.tsx\"",
```

**Review by jho406 — COMMENTED** [review]
Just the last those last 2 comments. I looked through the PR. CI is failing because of eslint. We can ignore that, i'll add another story for that. The output is also slightly different because typescript caught an undefined export `updateFragments`, which is great!

I had to other comments that i'd like addressed, and i'll merge away! Thanks for working on this.

File: superglue/package.json (hunk 5 above)
Comment (johncardiologs) [reply superglue/package.json]: Done!

**Pull request merged**